### ISSUE TITLE
Duration mismatching problem for WenetSpeech

### ISSUE THREAD

**Issue by pkufool** [lhotse-speech/lhotse#497]
As I mentioned here https://github.com/lhotse-speech/lhotse/pull/487#issuecomment-977652053

I always got duration mismatching problem when extracting features for WenetSpeech, the mismatching could be hundreds seconds. I checked the audio files, the real durations are normal.

![image](https://user-images.githubusercontent.com/11765074/144210252-cd80fcb8-d8fe-4119-8b66-679c4263c6a9.png)

I thought this is would be the audio data reading problem.

**Comment by pkufool** [comment]
Another one:

```
/ceph-hw/kangwei/dev_tools/anaconda3/envs/k4/lib/python3.8/site-packages/lhotse-1.0.0.dev0+git.5bf44d4.dirty-py3.8.egg/lhotse/dataset/sampling/single_cut.py:2
37: UserWarning: The first cut drawn in batch collection violates the max_frames, max_cuts, or max_duration constraints - we'll return it anyway. Consider inc
reasing max_frames/max_cuts/max_duration.                                                                                                                     
Computing features in batches:   0%|                                                                                  | 191/175665 [05:47<88:47:52,  1.82s/it]
Traceback (most recent call last):                                                                                                                            
  File "./local/compute_fbank_wenet_speech.py", line 123, in <module>                                                                                         
    compute_fbank_wenetspeech(args)                                                                                                                           
  File "./local/compute_fbank_wenet_speech.py", line 85, in compute_fbank_wenetspeech                                                                         
    cut_set = cut_set.compute_and_store_features_batch(                                                                                                       
  File "/ceph-hw/kangwei/dev_tools/anaconda3/envs/k4/lib/python3.8/site-packages/lhotse-1.0.0.dev0+git.5bf44d4.dirty-py3.8.egg/lhotse/cut.py", line 3819, in c
ompute_and_store_features_batch                                                                                                                               
  File "/ceph-hw/kangwei/dev_tools/anaconda3/envs/k4/lib/python3.8/site-packages/torch/utils/data/dataloader.py", line 435, in __next__                       
    data = self._next_data()                                                                                                                                  
  File "/ceph-hw/kangwei/dev_tools/anaconda3/envs/k4/lib/python3.8/site-packages/torch/utils/data/dataloader.py", line 1065, in _next_data                    
    return self._process_data(data)                                                                                                                           
  File "/ceph-hw/kangwei/dev_tools/anaconda3/envs/k4/lib/python3.8/site-packages/torch/utils/data/dataloader.py", line 1111, in _process_data                 
    data.reraise()                                                                                                                                            
  File "/ceph-hw/kangwei/dev_tools/anaconda3/envs/k4/lib/python3.8/site-packages/torch/_utils.py", line 428, in reraise                                       
    raise self.exc_type(msg)                                                                                                                                  
ValueError: Caught ValueError in DataLoader worker process 11.                                                                                                
Original Traceback (most recent call last):                                                                                                                   
  File "/ceph-hw/kangwei/dev_tools/anaconda3/envs/k4/lib/python3.8/site-packages/lhotse-1.0.0.dev0+git.5bf44d4.dirty-py3.8.egg/lhotse/utils.py", line 587, in 
wrapper                                                                                                                                                       
    return fn(*args, **kwargs)                                                                                                                                
  File "/ceph-hw/kangwei/dev_tools/anaconda3/envs/k4/lib/python3.8/site-packages/lhotse-1.0.0.dev0+git.5bf44d4.dirty-py3.8.egg/lhotse/audio.py", line 354, in 
load_audio                                                                                                                                                    
    samples = source.load_audio(                                                                                                                              
  File "/ceph-hw/kangwei/dev_tools/anaconda3/envs/k4/lib/python3.8/site-packages/lhotse-1.0.0.dev0+git.5bf44d4.dirty-py3.8.egg/lhotse/audio.py", line 152, in 
load_audio                                                                                                                                                    
    raise ValueError(                                                                                                                                         
ValueError: Requested more audio (2496.82s) than available (2239.999s)  
```

**Comment by pkufool** [comment]
Ah, from the error logs, it seems that the `ffmpeg` cmd run failed.
```
ValueError: Could not determine the number of channels for OPUS file from the following ffmpeg output (shown as bytestring due to avoid possible encoding issu
es):                                                                                                                                                          
b''                                                                                                                                                           
The ffmpeg command for which the program failed is: 'ffmpeg -threads 1 -t 3807.84 -i /ceph-hw/kangwei/code/icefall_wenet/egs/wenetspeech/ASR/download/wenetspe
ech/audio/train/youtube/B00000/Y0000000157_-P1DtPmLFYQ.opus -ar 16000 -f f32le -threads 1 pipe:1'
```
The returned `proc.stderr` is an empty byte array.

https://github.com/lhotse-speech/lhotse/blob/4301733a67430608f957c50b7ae724f7b8fbd18d/lhotse/audio.py#L1515-L1532

https://github.com/lhotse-speech/lhotse/blob/4301733a67430608f957c50b7ae724f7b8fbd18d/lhotse/audio.py#L1535-L1556

**Comment by pkufool** [comment]
I think the easiest thing to do is just continuing when the cmd fails.

**Comment by pzelasko** [comment]
I think we should debug this a little more, I don't understand why ffmpeg is failing. Can you try running the command that failed in the shell and see if there is any informative output?

But you're also right we should support skipping over problematic files in feature extraction. I can look into it later, or if you have time, please make a PR.

**Comment by pkufool** [comment]
I run the failed command, the logs is below. 
The output raw data is a 233MB file. 
```
(k4) kangwei@de-74279-k2-train-4-0809194600-d77fc9b88-4vzzq:/ceph-hw/kangwei/code/icefall_wenet/egs/wenetspeech/ASR/data$ ffmpeg -threads 1 -t 3807.84 -i /ceph-hw/kangwei/code/icefall_wenet/egs/wenetspeech/ASR/download/wenetspeech/audio/train/youtube/B00000/Y0000000157_-P1DtPmLFYQ.opus -ar 16000 -f f32le -threads 1 1>std.raw 2>err.txt
```
The err.txt
```
(k4) kangwei@de-74279-k2-train-4-0809194600-d77fc9b88-4vzzq:/ceph-hw/kangwei/code/icefall_wenet/egs/wenetspeech/ASR/data$ cat err.txt 
ffmpeg version 3.4.8-0ubuntu0.2 Copyright (c) 2000-2020 the FFmpeg developers
  built with gcc 7 (Ubuntu 7.5.0-3ubuntu1~18.04)
  configuration: --prefix=/usr --extra-version=0ubuntu0.2 --toolchain=hardened --libdir=/usr/lib/x86_64-linux-gnu --incdir=/usr/include/x86_64-linux-gnu --enable-gpl --disable-stripping --enable-avresample --enable-avisynth --enable-gnutls --enable-ladspa --enable-libass --enable-libbluray --enable-libbs2b --enable-libcaca --enable-libcdio --enable-libflite --enable-libfontconfig --enable-libfreetype --enable-libfribidi --enable-libgme --enable-libgsm --enable-libmp3lame --enable-libmysofa --enable-libopenjpeg --enable-libopenmpt --enable-libopus --enable-libpulse --enable-librubberband --enable-librsvg --enable-libshine --enable-libsnappy --enable-libsoxr --enable-libspeex --enable-libssh --enable-libtheora --enable-libtwolame --enable-libvorbis --enable-libvpx --enable-libwavpack --enable-libwebp --enable-libx265 --enable-libxml2 --enable-libxvid --enable-libzmq --enable-libzvbi --enable-omx --enable-openal --enable-opengl --enable-sdl2 --enable-libdc1394 --enable-libdrm --enable-libiec61883 --enable-chromaprint --enable-frei0r --enable-libopencv --enable-libx264 --enable-shared
  libavutil      55. 78.100 / 55. 78.100
  libavcodec     57.107.100 / 57.107.100
  libavformat    57. 83.100 / 57. 83.100
  libavdevice    57. 10.100 / 57. 10.100
  libavfilter     6.107.100 /  6.107.100
  libavresample   3.  7.  0 /  3.  7.  0
  libswscale      4.  8.100 /  4.  8.100
  libswresample   2.  9.100 /  2.  9.100
  libpostproc    54.  7.100 / 54.  7.100
Input #0, ogg, from '/ceph-hw/kangwei/code/icefall_wenet/egs/wenetspeech/ASR/download/wenetspeech/audio/train/youtube/B00000/Y0000000157_-P1DtPmLFYQ.opus':
  Duration: 01:03:27.85, start: 0.000000, bitrate: 32 kb/s
    Stream #0:0: Audio: opus, 48000 Hz, mono, fltp
    Metadata:
      ENCODER         : Lavc57.107.100 libopus
Stream mapping:
  Stream #0:0 -> #0:0 (opus (native) -> pcm_f32le (native))
Press [q] to stop, [?] for help
Output #0, f32le, to 'pipe:1':
  Metadata:
    encoder         : Lavf57.83.100
    Stream #0:0: Audio: pcm_f32le, 16000 Hz, mono, flt, 512 kb/s
    Metadata:
      encoder         : Lavc57.107.100 pcm_f32le
size=  237990kB time=01:03:27.83 bitrate= 512.0kbits/s speed= 390x    
video:0kB audio:237990kB subtitle:0kB other streams:0kB global headers:0kB muxing overhead: 0.000000%
```

Convert the raw data to a wave
```
$ sox -r 16k -e float -b 32 -c 1 std.raw tmp.wav
$ soxi tmp.wav
(k4) kangwei@de-74279-k2-train-4-0809194600-d77fc9b88-4vzzq:/ceph-hw/kangwei/code/icefall_wenet/egs/wenetspeech/ASR/data$ soxi tmp.wav

Input File     : 'tmp.wav'
Channels       : 1
Sample Rate    : 16000
Precision      : 25-bit
Duration       : 01:03:27.84 = 60925440 samples ~ 285588 CDDA sectors
File Size      : 244M
Bit Rate       : 512k
Sample Encoding: 32-bit Floating Point PCM
```
The wave is normal.

**Comment by pzelasko** [comment]
Catching an exception in the feature extraction loop, and skipping the problematic recordings. It can even be a dedicated exception type, eg AudioLoadongError (to replace ValueError in ffmpeg reading function)

**Comment by pkufool** [comment]
> Catching an exception in the feature extraction loop, and skipping the problematic recordings. It can even be a dedicated exception type, eg AudioLoadingError (to replace ValueError in ffmpeg reading function)

Thanks, catching exception is a good idea.

**Comment by danpovey** [comment]
Killed probably means they used up too much memory and OOM killer killed
them.

**Comment by csukuangfj** [comment]
Lhotse caches the most recent 512 audio files that have been read so far by default, which uses a lot of RAM if the audio files are very long.

**Comment by pzelasko** [comment]
These ones seem short, ffmpeg has args for offset and duration in his commands. But maybe the default cache is too large anyway..

**Comment by danpovey** [comment]
You're breaking thee manifest up into multiple pieces for processing, right?  Otherwise the manifest itself would take up too much memory.

**Comment by pkufool** [comment]
> You're breaking thee manifest up into multiple pieces for processing, right? Otherwise the manifest itself would take up too much memory.

I only tried S (100hours), M (1000hours) subset of the dataset, the manifests are not very big.

**Comment by pzelasko** [comment]
Is this resolved or is there anything else I can do to help?

**Comment by pkufool** [comment]
After changing to another machine(with faster IO), the duration mismatching error just happens in a very rare chances.  I created a pull request to skip the problematic audios just now, but I have not figured out how to catch exceptions in the multithead executor, it would be good if you give me some suggestions. see comments here #513

**Comment by pzelasko** [comment]
I think we resolved this with the `fault_tolerant` parameter merged in #533 -- LMK if this is good to close.

**Comment by pkufool** [comment]
@pzelasko  I think @luomingshuang  has verified this.  He has already extracted the features successfully.
